core/board.py

```diff
@@ -60,16 +62,46 @@ def layout(self) -> list[list[int]]:
     def enumerate(self) -> list[tuple[Hex, Color]]:
         """
         Returns all positions and values on the game board a la `enumerate`.
-        :return: a list of (Hex, T) tuples
+        :return: a list of (Hex, Color) tuples
         """
-        if not self._items:
-            self._items = []
+        if not self.__items:
+            items = []
             for r, line in enumerate(self._data):
                 for q, val in enumerate(line):
                     q += self.offset(r)
                     item = (Hex(q, r), val)
-                    self._items.append(item)
-        return self._items
+                    items.append(item)
+            self.__items = items
+
+        if self.__items_dirty:
+            self.__items_dirty = False
+            items = self.__items
+            i = 0
+            for r, line in enumerate(self._data):
+                for q, val in enumerate(line):
+                    q += self.offset(r)
+                    item = (Hex(q, r), val)
+                    items[i] = item
+                    i += 1
+
+        return self.__items
+
+    def enumerate_nonempty(self) -> list[tuple[Hex, Color]]:
+        """
+        Enumerates all non-empty (cell, value) pairs.
+        :return: a list of (Hex, Color) tuples
+        """
+        if not self.__items_nonempty:
+            items = []
+            for r, line in enumerate(self._data):
+                for q, val in enumerate(line):
+                    if val is None:
+                        continue
+                    q += self.offset(r)
+                    item = (Hex(q, r), val)
+                    items.append(item)
+            self.__items_nonempty = items
+        return self.__items_nonempty
```

Comment:
So this is an optimization that comes in handy a couple of times in heuristic calculation and zobrist hashing..... you'll notice that in a lot of places where `Board.enumerate` is used we just ignore empty cells. By caching non-empty cells we cut down at least 33 out of the 61 iterations we'd do with a standard enumerate call which actually does yield a non-trivial performance improvement. (it essentially reuses code to avoid branching overhead per cell, but we can't all get what we want)